### PR TITLE
[examples/summarization] deal with `max_length` and `num_beams`

### PR DESCRIPTION
**Pull request by bofenghuang**
# What does this PR do?

Hello 👋,

1. In the examples/pytorch/summarization/run_summarization.py, when the `generation_max_length` is not defined, the parameter `max_length` for `generate` function will be set to `val_max_target_length` in the line [675](https://github.com/huggingface/transformers/blob/82e61f34451dbea2de8d2220d51b0609d605dfd7/examples/pytorch/summarization/run_summarization.py#L675-L679), and be used for the **final evaluation** and prediction after training. However, the `max_length` used for the **evaluation during the training** will be possibly set to `None` or the `max_length` defined in the model's config. It's not a consistent behavior. Here I tried to unify this parameter before the `Seq2SeqTrainer` is initialized. Also applied the same to `num_beams`.

2. In the examples/pytorch/summarization/run_summarization_no_trainer.py, fixed the parameter `max_length`. Think it has higher priority than `val_max_target_length`.

Please ignore this PR if it's intended behavior :)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

cc: @sgugger @patil-suraj

### CODE REVIEW

File: examples/pytorch/summarization/run_summarization_no_trainer.py [hunk 1]
```diff
@@ -164,11 +164,8 @@ def parse_args():
     parser.add_argument(
         "--max_length",
         type=int,
-        default=128,
-        help=(
-            "The maximum total input sequence length after tokenization. Sequences longer than this will be truncated,"
-            " sequences shorter will be padded if `--pad_to_max_lengh` is passed."
-        ),
+        default=None,
+        help=("The maximum target length to use when predicting with the generate method."),
```
Comment (sgugger) [inline examples/pytorch/summarization/run_summarization_no_trainer.py:168]: This is the maximum length for the inputs, not the targets. I don't think this should be touched.

File: examples/pytorch/summarization/run_summarization_no_trainer.py [hunk 2]
```diff
@@ -671,7 +668,7 @@ def postprocess_text(preds, labels):
             args.val_max_target_length = args.max_target_length
 
         gen_kwargs = {
-            "max_length": args.val_max_target_length if args is not None else config.max_length,
+            "max_length": args.max_length if args.max_length is not None else args.val_max_target_length,
```
Comment (sgugger) [inline examples/pytorch/summarization/run_summarization_no_trainer.py:671]: It should be the `max_target_length` here if `val_max_target_length` is `None`, `max_length` is for the inputs only.

**Review by sgugger — COMMENTED** [review]
Thanks for your PR, I left a few comments.

File: examples/pytorch/summarization/run_summarization_no_trainer.py (hunk 1 above)
Comment (bofenghuang) [reply examples/pytorch/summarization/run_summarization_no_trainer.py:168]: There is another argument for the inputs `max_source_length`. So I suppose the `max_length` is used to constrain the generation length

Comment (sgugger) [reply examples/pytorch/summarization/run_summarization_no_trainer.py:168]: It's actually never used if I look more closely, so let's remove it entirely.

File: examples/pytorch/summarization/run_summarization_no_trainer.py (hunk 2 above)
Comment (sgugger) [reply examples/pytorch/summarization/run_summarization_no_trainer.py:671]: So you can just use `max_target_length` here or the config max_length if it was set to None.

Comment (bofenghuang) [reply examples/pytorch/summarization/run_summarization_no_trainer.py:671]: `max_target_length` is also used in `preprocess_function` to truncate the outputs. 

Here I think it might be better to provide another generation max length parameter, which is only used for evaluation/prediction, as done in examples/pytorch/summarization/run_summarization.py. In case the users don't want to truncate the summaries in dataset, but want to limit the generation max length in evaluation to accelerate the whole training process?

And of course if it is `None`, it will be set to the `val_max_target_length` (or `max_target_length`) or the `max_length` in model's config

Comment (sgugger) [reply examples/pytorch/summarization/run_summarization_no_trainer.py:671]: I don't think so. The parameter comes from a copy-paste and wasn't used, I wouldn't find a use for it. If the user wants some none-standard behavior, they should adapt the example.

**Comment by bofenghuang** [comment]
Hi @sgugger, just removed the `max_length`!

I also added another modif which separates the preprocessing of the training and the validation set in order to truncate the outputs by `max_target_length` and `val_max_target_length` respectively

**Review by sgugger — APPROVED** [review]
Thanks for iterating!
